File: runtime/core/federation/consensus.py
# core/federation/consensus.py - Byzantine Fault Tolerant Pattern Consensus
#
# Implements 3-of-5 (or N-of-M) Byzantine consensus for pattern acceptance.
# Based on PBFT (Practical Byzantine Fault Tolerance) simplified for patterns.
#
# Consensus Flow:
# 1. PRE-PREPARE: Leader proposes pattern for consensus
# 2. PREPARE: Validators verify and broadcast prepare messages
# 3. COMMIT: Upon 2f+1 prepares, broadcast commit
# 4. FINALIZE: Upon 2f+1 commits, pattern is accepted

from __future__ import annotations

import logging
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Callable, Dict, List, Optional, Set

from core.federation.protocol import (
    ConsensusVote,
    ConsensusResult,
    PatternEnvelope,
    VoteDecision,
    MIN_IHSAN_SCORE,
    canonical_json,
    sign_message,
)

logger = logging.getLogger("federation.consensus")


# ═══════════════════════════════════════════════════════════════════════════════
# CONSTANTS
# ═══════════════════════════════════════════════════════════════════════════════

# Byzantine parameters: N = 5, f = 1, quorum = 2f + 1 = 3
CONSENSUS_N = 5  # Total validators
CONSENSUS_F = 1  # Max Byzantine failures
CONSENSUS_QUORUM = 2 * CONSENSUS_F + 1  # = 3

# Timeouts
CONSENSUS_TIMEOUT_SEC = 30.0  # Max time for consensus round
VOTE_TIMEOUT_SEC = 10.0  # Time to wait for votes


# ═══════════════════════════════════════════════════════════════════════════════
# CONSENSUS STATE
# ═══════════════════════════════════════════════════════════════════════════════


class ConsensusPhase(Enum):
    """Phases of consensus."""

    IDLE = "idle"
    PRE_PREPARE = "pre_prepare"
    PREPARE = "prepare"
    COMMIT = "commit"
    FINALIZED = "finalized"
    FAILED = "failed"


@dataclass
class ConsensusState:
    """State of an ongoing consensus round."""

    pattern_id: str
    phase: ConsensusPhase = ConsensusPhase.IDLE
    started_at: float = field(default_factory=time.time)

    # Votes collected
    prepare_votes: Dict[str, ConsensusVote] = field(default_factory=dict)
    commit_votes: Dict[str, ConsensusVote] = field(default_factory=dict)

    # Result
    result: Optional[ConsensusResult] = None

    def is_expired(self) -> bool:
        """Check if consensus round has timed out."""
        return time.time() - self.started_at > CONSENSUS_TIMEOUT_SEC

    def prepare_count(self) -> int:
        """Count ACCEPT votes in prepare phase."""
        return sum(
            1 for v in self.prepare_votes.values() if v.decision == VoteDecision.ACCEPT
        )

    def commit_count(self) -> int:
        """Count ACCEPT votes in commit phase."""
        return sum(
            1 for v in self.commit_votes.values() if v.decision == VoteDecision.ACCEPT
        )

    def has_prepare_quorum(self) -> bool:
        """Check if prepare quorum reached."""
        return self.prepare_count() >= CONSENSUS_QUORUM

    def has_commit_quorum(self) -> bool:
        """Check if commit quorum reached."""
        return self.commit_count() >= CONSENSUS_QUORUM


# ═══════════════════════════════════════════════════════════════════════════════
# PATTERN VALIDATOR
# ═══════════════════════════════════════════════════════════════════════════════


class PatternValidator:
    """
    Validates patterns for consensus.

    Checks:
    1. Signature validity (cryptographic)
    2. Ihsān score meets threshold
    3. Pattern structure is valid
    4. No known malicious patterns
    5. Node reputation check
    """

    def __init__(self):
        # Known malicious pattern hashes
        self.blacklisted_patterns: Set[str] = set()

        # Node reputation cache
        self.node_reputation: Dict[str, float] = {}

        # Validation history
        self.validation_history: List[Dict[str, Any]] = []

    def validate(
        self,
        envelope: PatternEnvelope,
        local_ihsan_score: Optional[float] = None,
    ) -> tuple[VoteDecision, str, float]:
        """
        Validate pattern and return vote decision.

        Returns:
            (decision, reason, computed_ihsan)
        """
        # 1. Check blacklist
        if envelope.metadata.pattern_id in self.blacklisted_patterns:
            return VoteDecision.REJECT, "Pattern blacklisted", 0.0

        # 2. Verify cryptographic integrity
        valid, reason = envelope.verify()
        if not valid:
            return VoteDecision.REJECT, f"Verification failed: {reason}", 0.0

        # 3. Check Ihsān score
        ihsan = envelope.metadata.ihsan_score
        if ihsan < MIN_IHSAN_SCORE:
            return VoteDecision.REJECT, f"Ihsān {ihsan:.2f} < {MIN_IHSAN_SCORE}", ihsan

        # 4. Check node reputation
        origin = envelope.metadata.origin_node_id
        if origin in self.node_reputation:
            if self.node_reputation[origin] < 0.5:
                return VoteDecision.REJECT, "Origin node reputation too low", ihsan

        # 5. Check impact score
        if envelope.metadata.impact_score < 0.5:
            return (
                VoteDecision.REJECT,
                f"Impact score too low: {envelope.metadata.impact_score}",
                ihsan,
            )

        # 6. Check repetition count
        if envelope.metadata.repetition_count < 3:
            return VoteDecision.REJECT, "Insufficient repetitions", ihsan

        # 7. If we have local evaluation, weight it
        if local_ihsan_score is not None:
            # Combine origin and local scores
            combined = 0.7 * ihsan + 0.3 * local_ihsan_score
            if combined < MIN_IHSAN_SCORE:
                return (
                    VoteDecision.REJECT,
                    f"Combined Ihsān {combined:.2f} too low",
                    combined,
                )
            ihsan = combined

        # All checks passed
        return VoteDecision.ACCEPT, "Valid pattern", ihsan

    def update_reputation(self, node_id: str, delta: float):
        """Update node reputation."""
        current = self.node_reputation.get(node_id, 1.0)
        new_rep = max(0.0, min(1.0, current + delta))
        self.node_reputation[node_id] = new_rep

    def blacklist_pattern(self, pattern_id: str):
        """Add pattern to blacklist."""
        self.blacklisted_patterns.add(pattern_id)


# ═══════════════════════════════════════════════════════════════════════════════
# CONSENSUS ENGINE
# ═══════════════════════════════════════════════════════════════════════════════


class PatternConsensus:
    """
    Byzantine Fault Tolerant consensus for pattern acceptance.

    Uses simplified PBFT with 3-of-5 quorum requirement.
    Ensures network-wide agreement on pattern validity.
    """

    def __init__(
        self,
        node_id: str,
        private_key: bytes,
        public_key: bytes,
    ):
        self.node_id = node_id
        self._private_key = private_key
        self._public_key = public_key

        # Active consensus rounds
        self.active_rounds: Dict[str, ConsensusState] = {}

        # Completed rounds (recent)
        self.completed_rounds: Dict[str, ConsensusResult] = {}

        # Validator
        self.validator = PatternValidator()

        # Callbacks
        self._on_pattern_accepted: Optional[Callable[[PatternEnvelope], None]] = None
        self._on_pattern_rejected: Optional[Callable[[str, str], None]] = None
        self._broadcast_vote: Optional[Callable[[ConsensusVote], None]] = None

        # Statistics
        self.stats = ConsensusStats()

    # ─────────────────────────────────────────────────────────────────────────
    # Consensus Operations
    # ─────────────────────────────────────────────────────────────────────────

    async def propose_pattern(self, envelope: PatternEnvelope) -> ConsensusState:
        """
        Propose a new pattern for consensus (PRE-PREPARE phase).
        """
        pattern_id = envelope.metadata.pattern_id

        # Already in consensus?
        if pattern_id in self.active_rounds:
            return self.active_rounds[pattern_id]

        if pattern_id in self.completed_rounds:
            # Already decided
            state = ConsensusState(
                pattern_id=pattern_id, phase=ConsensusPhase.FINALIZED
            )
            state.result = self.completed_rounds[pattern_id]
            return state

        # Create new consensus round
        state = ConsensusState(pattern_id=pattern_id, phase=ConsensusPhase.PRE_PREPARE)
        self.active_rounds[pattern_id] = state

        logger.info(f"📋 Starting consensus for pattern {pattern_id[:16]}...")

        # Validate and cast our vote
        decision, reason, ihsan = self.validator.validate(envelope)

        our_vote = self._create_vote(pattern_id, decision, reason, ihsan)
        state.prepare_votes[self.node_id] = our_vote

        # Move to PREPARE phase
        state.phase = ConsensusPhase.PREPARE

        # Broadcast vote
        if self._broadcast_vote:
            self._broadcast_vote(our_vote)

        self.stats.rounds_started += 1

        return state

    async def receive_vote(
        self,
        vote: ConsensusVote,
        phase: str = "prepare",
    ) -> Optional[ConsensusResult]:
        """
        Receive a vote from another validator.

        Returns ConsensusResult if finalized.
        """
        pattern_id = vote.pattern_id

        # Already completed?
        if pattern_id in self.completed_rounds:
            return self.completed_rounds[pattern_id]

        # Get or create state
        if pattern_id not in self.active_rounds:
            self.active_rounds[pattern_id] = ConsensusState(
                pattern_id=pattern_id,
                phase=ConsensusPhase.PREPARE,
            )

        state = self.active_rounds[pattern_id]

        # Check timeout
        if state.is_expired():
            return self._finalize_timeout(state)

        # Record vote
        if phase == "prepare":
            state.prepare_votes[vote.voter_id] = vote

            # Check for prepare quorum
            if state.has_prepare_quorum() and state.phase == ConsensusPhase.PREPARE:
                state.phase = ConsensusPhase.COMMIT
                logger.debug(f"Pattern {pattern_id[:16]} reached prepare quorum")

                # Cast commit vote
                commit_vote = self._create_vote(
                    pattern_id,
                    VoteDecision.ACCEPT,
                    "Prepare quorum reached",
                    vote.ihsan_score,
                )
                state.commit_votes[self.node_id] = commit_vote

                if self._broadcast_vote:
                    self._broadcast_vote(commit_vote)

        elif phase == "commit":
            state.commit_votes[vote.voter_id] = vote

            # Check for commit quorum
            if state.has_commit_quorum() and state.phase == ConsensusPhase.COMMIT:
                return self._finalize_accept(state)

        return None

    def _create_vote(
        self,
        pattern_id: str,
        decision: VoteDecision,
        reason: str,
        ihsan: float,
    ) -> ConsensusVote:
        """Create a signed vote."""
        timestamp = datetime.now(timezone.utc).isoformat()

        # Create signature content
        content = canonical_json(
            {
                "pattern_id": pattern_id,
                "voter_id": self.node_id,
                "decision": decision.value,
                "reason": reason,
                "ihsan_score": ihsan,
                "timestamp": timestamp,
            }
        )

        signature = sign_message(content.encode(), self._private_key)

        return ConsensusVote(
            pattern_id=pattern_id,
            voter_id=self.node_id,
            decision=decision,
            reason=reason,
            ihsan_score=ihsan,
            timestamp=timestamp,
            signature=signature.hex(),
        )

    def _finalize_accept(self, state: ConsensusState) -> ConsensusResult:
        """Finalize consensus as accepted."""
        state.phase = ConsensusPhase.FINALIZED

        all_votes = list(state.prepare_votes.values()) + list(
            state.commit_votes.values()
        )

        result = ConsensusResult(
            pattern_id=state.pattern_id,
            accepted=True,
            accept_votes=state.commit_count(),
            reject_votes=sum(
                1
                for v in state.commit_votes.values()
                if v.decision == VoteDecision.REJECT
            ),
            abstain_votes=sum(
                1
                for v in state.commit_votes.values()
                if v.decision == VoteDecision.ABSTAIN
            ),
            quorum_reached=True,
            finalized_at=datetime.now(timezone.utc).isoformat(),
            votes=all_votes,
        )

        state.result = result
        self.completed_rounds[state.pattern_id] = result
        del self.active_rounds[state.pattern_id]

        self.stats.rounds_accepted += 1

        logger.info(
            f"✅ Pattern {state.pattern_id[:16]} ACCEPTED by consensus ({result.accept_votes}/{CONSENSUS_QUORUM})"
        )

        if self._on_pattern_accepted:
            # Callback would receive the envelope
            pass

        return result

    def _finalize_timeout(self, state: ConsensusState) -> ConsensusResult:
        """Finalize consensus as failed due to timeout."""
        state.phase = ConsensusPhase.FAILED

        result = ConsensusResult(
            pattern_id=state.pattern_id,
            accepted=False,
            accept_votes=state.prepare_count(),
            reject_votes=sum(
                1
                for v in state.prepare_votes.values()
                if v.decision == VoteDecision.REJECT
            ),
            abstain_votes=sum(
                1
                for v in state.prepare_votes.values()
                if v.decision == VoteDecision.ABSTAIN
            ),
            quorum_reached=False,
            finalized_at=datetime.now(timezone.utc).isoformat(),
            votes=list(state.prepare_votes.values()),
        )

        state.result = result
        self.completed_rounds[state.pattern_id] = result
        del self.active_rounds[state.pattern_id]

        self.stats.rounds_timeout += 1

        logger.warning(
            f"⏱️ Pattern {state.pattern_id[:16]} TIMEOUT - no consensus reached"
        )

        return result

    # ─────────────────────────────────────────────────────────────────────────
    # Callbacks
    # ─────────────────────────────────────────────────────────────────────────

    def on_pattern_accepted(self, callback: Callable[[PatternEnvelope], None]):
        """Register callback for accepted patterns."""
        self._on_pattern_accepted = callback

    def on_pattern_rejected(self, callback: Callable[[str, str], None]):
        """Register callback for rejected patterns."""
        self._on_pattern_rejected = callback

    def set_vote_broadcaster(self, callback: Callable[[ConsensusVote], None]):
        """Set callback to broadcast votes to peers."""
        self._broadcast_vote = callback

    # ─────────────────────────────────────────────────────────────────────────
    # Query
    # ─────────────────────────────────────────────────────────────────────────

    def get_result(self, pattern_id: str) -> Optional[ConsensusResult]:
        """Get consensus result for a pattern."""
        return self.completed_rounds.get(pattern_id)

    def is_accepted(self, pattern_id: str) -> bool:
        """Check if pattern was accepted."""
        result = self.completed_rounds.get(pattern_id)
        return result is not None and result.accepted

    def cleanup_old_rounds(self, max_age_sec: float = 3600.0):
        """Clean up old completed rounds."""
        now = time.time()
        cutoff = datetime.fromtimestamp(now - max_age_sec, timezone.utc).isoformat()

        to_remove = [
            pid
            for pid, result in self.completed_rounds.items()
            if result.finalized_at < cutoff
        ]

        for pid in to_remove:
            del self.completed_rounds[pid]


@dataclass
class ConsensusStats:
    """Consensus statistics."""

    rounds_started: int = 0
    rounds_accepted: int = 0
    rounds_rejected: int = 0
    rounds_timeout: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "rounds_started": self.rounds_started,
            "rounds_accepted": self.rounds_accepted,
            "rounds_rejected": self.rounds_rejected,
            "rounds_timeout": self.rounds_timeout,
            "acceptance_rate": (self.rounds_accepted / max(1, self.rounds_started)),
        }

File: runtime/src/federation/consensus.rs
// src/federation/consensus.rs - Byzantine Fault Tolerant Pattern Consensus
//
// Implements 3-of-5 (or N-of-M) Byzantine consensus for pattern acceptance.
// Compatible with Python implementation in core/federation/consensus.py

use crate::federation::protocol::{
    verify_signature, ConsensusResult, ConsensusVote, PatternEnvelope, SignedVote, VoteDecision,
    MIN_IHSAN_SCORE,
};
use anyhow::Result;
use chrono::{DateTime, Utc};
use ed25519_dalek::{Signer, SigningKey, VerifyingKey};
use std::collections::HashMap;
use std::time::{Duration, Instant};
use tracing::{debug, info, warn};

/// Byzantine parameters: N = 5, f = 1, quorum = 2f + 1 = 3
pub const CONSENSUS_N: usize = 5;
pub const CONSENSUS_F: usize = 1;
pub const CONSENSUS_QUORUM: usize = 2 * CONSENSUS_F + 1; // = 3

/// Consensus timeout in seconds
pub const CONSENSUS_TIMEOUT_SEC: u64 = 30;

// ═══════════════════════════════════════════════════════════════════════════════
// CONSENSUS STATE
// ═══════════════════════════════════════════════════════════════════════════════

/// Phases of consensus
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConsensusPhase {
    Idle,
    PrePrepare,
    Prepare,
    Commit,
    Finalized,
    Failed,
}

/// State of an ongoing consensus round
#[derive(Debug, Clone)]
pub struct ConsensusState {
    pub pattern_id: String,
    pub phase: ConsensusPhase,
    pub started_at: Instant,

    /// Votes collected in prepare phase
    pub prepare_votes: HashMap<String, ConsensusVote>,
    /// Votes collected in commit phase
    pub commit_votes: HashMap<String, ConsensusVote>,

    /// Final result
    pub result: Option<ConsensusResult>,
}

impl ConsensusState {
    pub fn new(pattern_id: String) -> Self {
        Self {
            pattern_id,
            phase: ConsensusPhase::Idle,
            started_at: Instant::now(),
            prepare_votes: HashMap::new(),
            commit_votes: HashMap::new(),
            result: None,
        }
    }

    pub fn is_expired(&self) -> bool {
        self.started_at.elapsed() > Duration::from_secs(CONSENSUS_TIMEOUT_SEC)
    }

    pub fn prepare_count(&self) -> usize {
        self.prepare_votes
            .values()
            .filter(|v| v.decision == VoteDecision::Accept)
            .count()
    }

    pub fn commit_count(&self) -> usize {
        self.commit_votes
            .values()
            .filter(|v| v.decision == VoteDecision::Accept)
            .count()
    }

    pub fn has_prepare_quorum(&self) -> bool {
        self.prepare_count() >= CONSENSUS_QUORUM
    }

    pub fn has_commit_quorum(&self) -> bool {
        self.commit_count() >= CONSENSUS_QUORUM
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// PATTERN VALIDATOR
// ═══════════════════════════════════════════════════════════════════════════════

/// Validates patterns for consensus
pub struct PatternValidator {
    /// Known malicious pattern hashes
    blacklisted_patterns: std::collections::HashSet<String>,
    /// Node reputation cache
    node_reputation: HashMap<String, f64>,
}

impl PatternValidator {
    pub fn new() -> Self {
        Self {
            blacklisted_patterns: std::collections::HashSet::new(),
            node_reputation: HashMap::new(),
        }
    }

    /// Validate pattern and return vote decision
    pub fn validate(
        &self,
        envelope: &PatternEnvelope,
        local_ihsan_score: Option<f64>,
    ) -> (VoteDecision, String, f64) {
        // 1. Check blacklist
        if self
            .blacklisted_patterns
            .contains(&envelope.metadata.pattern_id)
        {
            return (VoteDecision::Reject, "Pattern blacklisted".to_string(), 0.0);
        }

        // 2. Verify cryptographic integrity
        match envelope.verify() {
            Ok(true) => {}
            Ok(false) => {
                return (VoteDecision::Reject, "Verification failed".to_string(), 0.0);
            }
            Err(e) => {
                return (
                    VoteDecision::Reject,
                    format!("Verification error: {}", e),
                    0.0,
                );
            }
        }

        // 3. Check Ihsān score
        let ihsan = envelope.metadata.ihsan_score;
        if ihsan < MIN_IHSAN_SCORE {
            return (
                VoteDecision::Reject,
                format!("Ihsān {:.2} < {:.2}", ihsan, MIN_IHSAN_SCORE),
                ihsan,
            );
        }

        // 4. Check node reputation
        let origin = &envelope.metadata.origin_node_id;
        if let Some(&rep) = self.node_reputation.get(origin) {
            if rep < 0.5 {
                return (
                    VoteDecision::Reject,
                    "Origin node reputation too low".to_string(),
                    ihsan,
                );
            }
        }

        // 5. Check impact score
        if envelope.metadata.impact_score < 0.5 {
            return (
                VoteDecision::Reject,
                format!(
                    "Impact score too low: {:.2}",
                    envelope.metadata.impact_score
                ),
                ihsan,
            );
        }

        // 6. Check repetition count
        if envelope.metadata.repetition_count < 3 {
            return (
                VoteDecision::Reject,
                "Insufficient repetitions".to_string(),
                ihsan,
            );
        }

        // 7. Combine with local evaluation if provided
        let final_ihsan = if let Some(local) = local_ihsan_score {
            let combined = 0.7 * ihsan + 0.3 * local;
            if combined < MIN_IHSAN_SCORE {
                return (
                    VoteDecision::Reject,
                    format!("Combined Ihsān {:.2} too low", combined),
                    combined,
                );
            }
            combined
        } else {
            ihsan
        };

        (
            VoteDecision::Accept,
            "Valid pattern".to_string(),
            final_ihsan,
        )
    }

    /// Update node reputation
    pub fn update_reputation(&mut self, node_id: &str, delta: f64) {
        let current = self.node_reputation.get(node_id).copied().unwrap_or(1.0);
        let new_rep = (current + delta).clamp(0.0, 1.0);
        self.node_reputation.insert(node_id.to_string(), new_rep);
    }

    /// Blacklist a pattern
    pub fn blacklist_pattern(&mut self, pattern_id: &str) {
        self.blacklisted_patterns.insert(pattern_id.to_string());
    }
}

impl Default for PatternValidator {
    fn default() -> Self {
        Self::new()
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// CONSENSUS ENGINE
// ═══════════════════════════════════════════════════════════════════════════════

/// Consensus statistics
#[derive(Debug, Default, Clone)]
pub struct ConsensusStats {
    pub rounds_started: u64,
    pub rounds_accepted: u64,
    pub rounds_rejected: u64,
    pub rounds_timeout: u64,
}

impl ConsensusStats {
    pub fn acceptance_rate(&self) -> f64 {
        if self.rounds_started == 0 {
            0.0
        } else {
            self.rounds_accepted as f64 / self.rounds_started as f64
        }
    }
}

/// Byzantine Fault Tolerant consensus for pattern acceptance
pub struct PatternConsensus {
    node_id: String,
    private_key: Vec<u8>,
    public_key: Vec<u8>,

    /// Active consensus rounds
    active_rounds: HashMap<String, ConsensusState>,
    /// Completed rounds (recent)
    completed_rounds: HashMap<String, ConsensusResult>,
    /// Validator
    validator: PatternValidator,
    /// Statistics
    pub stats: ConsensusStats,
}

impl PatternConsensus {
    pub fn new(node_id: String, private_key: Vec<u8>, public_key: Vec<u8>) -> Self {
        Self {
            node_id,
            private_key,
            public_key,
            active_rounds: HashMap::new(),
            completed_rounds: HashMap::new(),
            validator: PatternValidator::new(),
            stats: ConsensusStats::default(),
        }
    }

    /// Propose a new pattern for consensus (PRE-PREPARE phase)
    pub async fn propose_pattern(&mut self, envelope: &PatternEnvelope) -> Result<ConsensusState> {
        let pattern_id = envelope.metadata.pattern_id.clone();

        // Already in consensus?
        if let Some(state) = self.active_rounds.get(&pattern_id) {
            return Ok(state.clone());
        }

        // Already completed?
        if let Some(result) = self.completed_rounds.get(&pattern_id) {
            let mut state = ConsensusState::new(pattern_id);
            state.phase = ConsensusPhase::Finalized;
            state.result = Some(result.clone());
            return Ok(state);
        }

        // Create new consensus round
        let mut state = ConsensusState::new(pattern_id.clone());
        state.phase = ConsensusPhase::PrePrepare;

        info!(
            "📋 Starting consensus for pattern {}...",
            &pattern_id[..16.min(pattern_id.len())]
        );

        // Validate and cast our vote
        let (decision, reason, ihsan) = self.validator.validate(envelope, None);

        let our_vote = self.create_vote(&pattern_id, decision, &reason, ihsan)?;
        state.prepare_votes.insert(self.node_id.clone(), our_vote);

        // Move to PREPARE phase
        state.phase = ConsensusPhase::Prepare;

        self.active_rounds.insert(pattern_id, state.clone());
        self.stats.rounds_started += 1;

        Ok(state)
    }

    /// Receive a vote from another validator
    pub async fn receive_vote(
        &mut self,
        vote: ConsensusVote,
        phase: &str,
    ) -> Option<ConsensusResult> {
        let pattern_id = vote.pattern_id.clone();

        // Already completed?
        if let Some(result) = self.completed_rounds.get(&pattern_id) {
            return Some(result.clone());
        }

        // Check for timeout first
        let is_expired = self
            .active_rounds
            .get(&pattern_id)
            .map(|s| s.is_expired())
            .unwrap_or(false);

        if is_expired {
            return Some(self.finalize_timeout(&pattern_id));
        }

        // Process the vote and determine next action
        let (need_commit_vote, ihsan_for_commit, should_finalize_accept, should_finalize_commit) = {
            // Get or create state
            let state = self
                .active_rounds
                .entry(pattern_id.clone())
                .or_insert_with(|| {
                    let mut s = ConsensusState::new(pattern_id.clone());
                    s.phase = ConsensusPhase::Prepare;
                    s
                });

            let mut need_commit = false;
            let mut ihsan = 0.0;
            let mut finalize_accept = false;
            let mut finalize_commit = false;

            // Record vote
            match phase {
                "prepare" => {
                    state
                        .prepare_votes
                        .insert(vote.voter_id.clone(), vote.clone());

                    // Check for prepare quorum
                    if state.has_prepare_quorum() && state.phase == ConsensusPhase::Prepare {
                        state.phase = ConsensusPhase::Commit;
                        debug!(
                            "Pattern {} reached prepare quorum",
                            &pattern_id[..16.min(pattern_id.len())]
                        );
                        need_commit = true;
                        ihsan = vote.ihsan_score;
                    }
                }
                "commit" => {
                    state.commit_votes.insert(vote.voter_id.clone(), vote);

                    // Check for commit quorum
                    if state.has_commit_quorum() && state.phase == ConsensusPhase::Commit {
                        finalize_commit = true;
                    }
                }
                _ => {}
            }

            (need_commit, ihsan, finalize_accept, finalize_commit)
        };

        // Now perform actions outside the borrow
        if need_commit_vote {
            if let Ok(commit_vote) = self.create_vote(
                &pattern_id,
                VoteDecision::Accept,
                "Prepare quorum reached",
                ihsan_for_commit,
            ) {
                if let Some(state) = self.active_rounds.get_mut(&pattern_id) {
                    state.commit_votes.insert(self.node_id.clone(), commit_vote);
                }
            }
        }

        if should_finalize_commit {
            return Some(self.finalize_accept(&pattern_id));
        }

        None
    }

    /// Receive a signed vote with Ed25519 verification (CRIT-1)
    ///
    /// Genesis Strict Synthesis v2.2.2: All votes MUST be verified before counting.
    /// This is the preferred method for receiving votes from other nodes.
    pub async fn receive_signed_vote(
        &mut self,
        signed_vote: SignedVote,
        phase: &str,
    ) -> Option<ConsensusResult> {
        // CRIT-1: Verify signature BEFORE processing
        if !signed_vote.verify() {
            warn!(
                "⚠️ REJECTED vote from {}: Ed25519 signature verification FAILED",
                signed_vote.vote.voter_id
            );
            return None;
        }

        debug!(
            "✓ Vote from {} verified (pattern: {})",
            signed_vote.vote.voter_id,
            &signed_vote.vote.pattern_id[..16.min(signed_vote.vote.pattern_id.len())]
        );

        // Extract verified vote and process
        self.receive_vote(signed_vote.into_vote(), phase).await
    }

    /// Create a signed vote
    fn create_vote(
        &self,
        pattern_id: &str,
        decision: VoteDecision,
        reason: &str,
        ihsan: f64,
    ) -> Result<ConsensusVote> {
        let timestamp = Utc::now();

        // Create signature content
        let content = serde_json::json!({
            "pattern_id": pattern_id,
            "voter_id": self.node_id,
            "decision": decision,
            "reason": reason,
            "ihsan_score": ihsan,
            "timestamp": timestamp.to_rfc3339(),
        });

        let content_str = serde_json::to_string(&content)?;

        // Sign (simplified - real implementation would use proper key handling)
        let private_key_bytes: [u8; 32] = self.private_key[..32]
            .try_into()
            .map_err(|_| anyhow::anyhow!("Invalid private key length"))?;
        let signing_key = SigningKey::from_bytes(&private_key_bytes);
        let signature = signing_key.sign(content_str.as_bytes());

        Ok(ConsensusVote {
            pattern_id: pattern_id.to_string(),
            voter_id: self.node_id.clone(),
            decision,
            reason: reason.to_string(),
            ihsan_score: ihsan,
            timestamp,
            signature: hex::encode(signature.to_bytes()),
        })
    }

    /// Create a signed vote (CRIT-1 compliant)
    ///
    /// Genesis Strict Synthesis v2.2.2: Returns SignedVote for transmission.
    /// Recipients MUST call receive_signed_vote() to verify before counting.
    pub fn create_signed_vote(
        &self,
        pattern_id: &str,
        decision: VoteDecision,
        reason: &str,
        ihsan: f64,
    ) -> Result<SignedVote> {
        let vote = self.create_vote(pattern_id, decision, reason, ihsan)?;

        // Get verifying key from private key
        let private_key_bytes: [u8; 32] = self.private_key[..32]
            .try_into()
            .map_err(|_| anyhow::anyhow!("Invalid private key length"))?;
        let signing_key = SigningKey::from_bytes(&private_key_bytes);
        let verifying_key = signing_key.verifying_key();

        Ok(SignedVote::from_vote(vote, &verifying_key))
    }

    /// Get our public key as hex
    pub fn public_key_hex(&self) -> String {
        hex::encode(&self.public_key)
    }

    /// Finalize consensus as accepted
    fn finalize_accept(&mut self, pattern_id: &str) -> ConsensusResult {
        let state = self.active_rounds.remove(pattern_id).unwrap();

        let all_votes: Vec<ConsensusVote> = state
            .prepare_votes
            .values()
            .chain(state.commit_votes.values())
            .cloned()
            .collect();

        let result = ConsensusResult {
            pattern_id: pattern_id.to_string(),
            accepted: true,
            accept_votes: state.commit_count(),
            reject_votes: state
                .commit_votes
                .values()
                .filter(|v| v.decision == VoteDecision::Reject)
                .count(),
            abstain_votes: state
                .commit_votes
                .values()
                .filter(|v| v.decision == VoteDecision::Abstain)
                .count(),
            quorum_reached: true,
            finalized_at: Utc::now(),
            votes: all_votes,
        };

        self.completed_rounds
            .insert(pattern_id.to_string(), result.clone());
        self.stats.rounds_accepted += 1;

        info!(
            "✅ Pattern {} ACCEPTED by consensus ({}/{})",
            &pattern_id[..16.min(pattern_id.len())],
            result.accept_votes,
            CONSENSUS_QUORUM
        );

        result
    }

    /// Finalize consensus as failed due to timeout
    fn finalize_timeout(&mut self, pattern_id: &str) -> ConsensusResult {
        let state = self.active_rounds.remove(pattern_id).unwrap();

        let result = ConsensusResult {
            pattern_id: pattern_id.to_string(),
            accepted: false,
            accept_votes: state.prepare_count(),
            reject_votes: state
                .prepare_votes
                .values()
                .filter(|v| v.decision == VoteDecision::Reject)
                .count(),
            abstain_votes: state
                .prepare_votes
                .values()
                .filter(|v| v.decision == VoteDecision::Abstain)
                .count(),
            quorum_reached: false,
            finalized_at: Utc::now(),
            votes: state.prepare_votes.values().cloned().collect(),
        };

        self.completed_rounds
            .insert(pattern_id.to_string(), result.clone());
        self.stats.rounds_timeout += 1;

        warn!(
            "⏱️ Pattern {} TIMEOUT - no consensus reached",
            &pattern_id[..16.min(pattern_id.len())]
        );

        result
    }

    /// Get consensus result for a pattern
    pub fn get_result(&self, pattern_id: &str) -> Option<&ConsensusResult> {
        self.completed_rounds.get(pattern_id)
    }

    /// Check if pattern was accepted
    pub fn is_accepted(&self, pattern_id: &str) -> bool {
        self.completed_rounds
            .get(pattern_id)
            .map(|r| r.accepted)
            .unwrap_or(false)
    }

    /// Clean up old completed rounds
    pub fn cleanup_old_rounds(&mut self, max_age: Duration) {
        let cutoff = Utc::now() - chrono::Duration::from_std(max_age).unwrap();

        self.completed_rounds
            .retain(|_, result| result.finalized_at > cutoff);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::federation::protocol::{PatternMetadata, PatternPayload, PatternType};

    #[test]
    fn test_consensus_state() {
        let state = ConsensusState::new("test_pattern".to_string());

        assert_eq!(state.phase, ConsensusPhase::Idle);
        assert!(!state.is_expired());
        assert!(!state.has_prepare_quorum());
        assert!(!state.has_commit_quorum());
    }

    #[test]
    fn test_pattern_validator() {
        let validator = PatternValidator::new();

        // Create test envelope
        let (private_key, public_key) = crate::federation::protocol::generate_keypair();

        let metadata = PatternMetadata {
            pattern_id: "test".to_string(),
            pattern_type: PatternType::SapeProbe,
            version: 1,
            origin_node_id: "node".to_string(),
            origin_timestamp: Utc::now(),
            repetition_count: 5,
            success_rate: 0.9,
            impact_score: 0.8,
            ihsan_score: 0.92,
            adoption_count: 0,
            expires_at: Utc::now() + chrono::Duration::days(30),
            tags: vec![],
        };

        let payload = PatternPayload {
            trigger_sequence: vec!["test".to_string()],
            optimization: "test".to_string(),
            latency_reduction_ms: 50,
            token_savings_percent: 20.0,
            snr_improvement: 0.1,
        };

        let envelope = crate::federation::protocol::PatternEnvelope::create(
            metadata,
            payload,
            &private_key,
            &public_key,
        )
        .unwrap();

        let (decision, _, _) = validator.validate(&envelope, None);
        assert_eq!(decision, VoteDecision::Accept);
    }

    #[test]
    fn test_consensus_stats() {
        let mut stats = ConsensusStats::default();

        assert_eq!(stats.acceptance_rate(), 0.0);

        stats.rounds_started = 10;
        stats.rounds_accepted = 8;

        assert!((stats.acceptance_rate() - 0.8).abs() < 0.001);
    }
}
